src/utils/unique_stack.rs

```diff
@@ -0,0 +1,221 @@
+    elements: Vec<Option<El<T>>>,
+    indices: HashMap<T, usize>,
+    top: Option<usize>,
+    free_indices: Vec<usize>,
+}
+
+impl<T: Eq + Hash + Clone> UniqueStack<T> {
+    /// Create a new empty UniqueStack.
+    pub fn new() -> Self {
+        UniqueStack {
+            elements: Vec::new(),
+            indices: HashMap::default(),
+            top: None,
+            free_indices: Vec::new(),
+        }
+    }
+
+    /// Push an element onto the stack, moving it to the top if already present.
+    ///
+    /// Return true if the element was added, false if it was already present.
+    pub fn push(&mut self, value: T) -> bool {
+        if let Some(&index) = self.indices.get(&value) {
+            self.rewire_top(index);
+            false
+        } else {
+            // New element, add it to the top
+            let new_index = self.get_free_index();
+            let el = El {
+                value: value.clone(),
+                below: None,
+                above: None,
+            };
+            self.elements[new_index] = Some(el);
+            self.indices.insert(value, new_index);
+            self.rewire_top(new_index);
+            true
+        }
+    }
+
+    /// Remove and return the top element from the stack.
+    ///
+    /// Return None if the stack is empty.
+    pub fn pop(&mut self) -> Option<T> {
+        let top_index = self.top?;
+        let el = self.elements[top_index].take().unwrap();
+        self.indices.remove(&el.value);
+        self.top = el.below;
+        Some(el.value)
+    }
+
+    /// The top element of the stack without removing it.
+    ///
+    /// Return None if the stack is empty.
+    pub fn top(&self) -> Option<&T> {
+        self.top
+            .map(|index| &self.elements[index].as_ref().unwrap().value)
+    }
+
+    /// A free spot in the vector, creating it if necessary
+    fn get_free_index(&mut self) -> usize {
+        self.free_indices.pop().unwrap_or_else(|| {
+            self.elements.push(None);
+            self.elements.len() - 1
+        })
+    }
+
+    /// Place index at top of the stack
+    fn rewire_top(&mut self, index: usize) {
+        let prev_top = self.top.replace(index);
+        if Some(index) == prev_top {
+            return;
+        }
+        let &El { below, above, .. } = self.elements[index].as_ref().unwrap();
+        self.link(below, above);
+        self.link(prev_top, Some(index));
+    }
+
+    /// Link two elements together by setting `above` and `below` indices
+    fn link(&mut self, below: Option<usize>, above: Option<usize>) {
+        if let Some(below) = below {
+            self.elements[below].as_mut().unwrap().above = above;
+        }
+        if let Some(above) = above {
+            self.elements[above].as_mut().unwrap().below = below;
+        }
+    }
+}
+
+impl<T: Eq + Hash + Clone> FromIterator<T> for UniqueStack<T> {
+    fn from_iter<I: IntoIterator<Item = T>>(iter: I) -> Self {
+        let mut stack = UniqueStack::new();
```

Comment:
Add a `UniqueStack::with_capacity` and pass `iter.size_hint().0` to it.

Reply:
Done!